bin_sort/bin_sort.c

```diff
@@ -0,0 +1,56 @@
+/* Following algorithm sorts the input array in ascending order
+ * Time Complexity : O(n)-> Avg case and O(n^2)-> Worst Case
+ */
+#include <stdio.h>
+#include <stdlib.h>
+
+struct bucket {
+	int count;
+	int* value;
+};
+
+int comparator(const int* first, const int* second) {
+	int a = *first, b = *second;
+	int result = a - b;
+	return (0 < result) - (result < 0);
+}
+
+void bucket_sort(int array[], int n) {
+	struct bucket buckets[3];
+	int a, b, c;
+	for (a = 0; a < 3; a++) {
+		buckets[a].count = 0;
+		buckets[a].value = malloc(sizeof(int) * n);
+	}
+	for (a = 0; a < n; a++) {
+		if (array[a] < 0) {
+			buckets[0].value[buckets[0].count++] = array[a];
+		}
+		else if (array[a] > 10) {
+			buckets[2].value[buckets[2].count++] = array[a];
+		}
+		else {
+			buckets[1].value[buckets[1].count++] = array[a];
+		}
+	}
+	for (c = 0, a = 0; a < 3; a++) {
+		qsort(buckets[a].value, buckets[a].count, sizeof(int), &comparator);
+		for (b = 0; b < buckets[a].count; b++) {
+			array[c + b] = buckets[a].value[b];
+		}
+		c += buckets[a].count;
+		free(buckets[a].value);
+	}
+}
+
+int main() {
+	int array[] = {-5, -9, 1000, 1, -10, 0, 2, 3, 5, 4, 1234, 7};
```

Comment:
From GFG:

> Bucket sort is mainly useful when input is uniformly distributed over a range. For example, consider the following problem. Sort a large set of floating point numbers which are in range from 0.0 to 1.0 and are uniformly distributed across the range. How do we sort the numbers efficiently?

This example does not reflects that. You should use uniform random numbers in the array. 
See the [Java implementation](https://github.com/iiitv/algos/blob/master/bin_sort/BinSort.java).